### PR TITLE
[WIP] Feature accept or reject results

### PR DESCRIPTION
**Pull request by rwrzesien**

### CODE REVIEW

File: concent_api/core/tests/test_integration_get_task_result.py [hunk 1]
```diff
@@ -208,6 +208,7 @@ def test_concent_requests_task_result_from_provider_and_requestor_receives_failu
         with freeze_time("2017-12-01 11:00:12"):
             response_2 = self.client.post(
                 reverse('core:receive'),
+                data = '',
```
Comment (rwrzesien) [inline concent_api/core/tests/test_integration_get_task_result.py]: Please remove this changes, they will come in different branch.

Comment (Jakub89) [reply concent_api/core/tests/test_integration_get_task_result.py]: Ok, I've deleted all [TMP] commits

File: concent_api/core/views.py [hunk 2]
```diff
@@ -93,6 +99,17 @@ def receive(request, _message):
         set_message_as_delivered(last_undelivered_message_status)
         return handle_receive_force_get_task_result_upload_for_provider(request, decoded_message_data)
 
+    if isinstance(decoded_message_data, message.concents.ForceSubtaskResults):
+        set_message_as_delivered(last_undelivered_message_status)
+        if current_time < decoded_message_data.timestamp + settings.CONCENT_MESSAGING_TIME:
```
Comment (rwrzesien) [inline concent_api/core/views.py]: According to https://github.com/golemfactory/concent/issues/44 we always want to return the message if it is queued.

File: concent_api/core/views.py [hunk 3]
```diff
@@ -93,6 +99,17 @@ def receive(request, _message):
...
+            return handle_receive_force_subtasks_results(decoded_message_data)
+        return None
+
+    if isinstance(decoded_message_data, message.concents.SubtaskResultsAccepted) or isinstance(decoded_message_data, message.concents.SubtaskResultsRejected):
```
Comment (rwrzesien) [inline concent_api/core/views.py]: You can pass a tuple of classes to `isinstance()` like this:
```
isinstance(decoded_message_data, (message.concents.SubtaskResultsAccepted, SubtaskResultsRejected))
```

File: concent_api/core/views.py [hunk 4]
```diff
@@ -271,6 +288,72 @@ def handle_send_force_get_task_result(client_message: message.concents.ForceGetT
...
+    assert client_message.TYPE in message.registered_message_types
+
+    current_time = int(datetime.datetime.now().timestamp())
+
+    if Message.objects.filter(
+        type    = client_message.TYPE,
+        task_id = client_message.ack_report_computed_task.task_to_compute.compute_task_def['task_id']
+    ).exists():
+        return message.concents.ServiceRefused(
+            timestamp   = client_message.timestamp,
```
Comment (rwrzesien) [inline concent_api/core/views.py]: We can't pass timestamp to message like this, please take a look on https://github.com/golemfactory/concent/pull/89/files, there are some ways to work around this.

File: concent_api/core/views.py [hunk 5]
```diff
@@ -271,6 +288,72 @@ def handle_send_force_get_task_result(client_message: message.concents.ForceGetT
...
+            reason      = message.concents.ServiceRefused.REASON.TaskAlreadyProcessed,
+        )
+
+    provider_has_enough_funds = get_provider_account_status(client_message)
+    if not provider_has_enough_funds:
+        return message.concents.ServiceRefused(
+            timestamp   = client_message.timestamp,
```
Comment (rwrzesien) [inline concent_api/core/views.py]: We can't pass timestamp to message like this, please take a look on https://github.com/golemfactory/concent/pull/89/files, there are some ways to work around this.

File: concent_api/core/views.py [hunk 6]
```diff
@@ -271,6 +288,72 @@ def handle_send_force_get_task_result(client_message: message.concents.ForceGetT
...
+    if Message.objects.filter(
+        type    = client_message.TYPE,
+        task_id = client_message.ack_report_computed_task.task_to_compute.compute_task_def['task_id']
+    ).exists():
+        return message.concents.ServiceRefused(
+            timestamp   = client_message.timestamp,
+            reason      = message.concents.ServiceRefused.REASON.TaskAlreadyProcessed,
+        )
+
+    provider_has_enough_funds = get_provider_account_status(client_message)
```
Comment (rwrzesien) [inline concent_api/core/views.py]: I know this is temporary, but two things here:
1. If method is returning True/False its name should start with `is_` or `has_`.
2. `provider_has_enough_funds` is not used anywhere later so you can call it directly in `if`.

File: concent_api/core/views.py [hunk 7]
```diff
@@ -271,6 +288,72 @@ def handle_send_force_get_task_result(client_message: message.concents.ForceGetT
...
+        return message.concents.ServiceRefused(
+            timestamp   = client_message.timestamp,
+            reason      = message.concents.ServiceRefused.REASON.TooSmallProviderDeposit,
+        )
+
+    client_message_send_to_soon = client_message.ack_report_computed_task['timestamp'] + settings.CONCENT_MESSAGING_TIME
+    client_message_send_to_late = client_message.ack_report_computed_task['timestamp'] + settings.CONCENT_MESSAGING_TIME + settings.FORCE_ACCEPTANCE_TIME
+    if client_message_send_to_late < current_time < client_message_send_to_soon:
+        return message.concents.ForceSubtaskResultsRejected(
+            timestamp   = client_message.timestamp,
```
Comment (rwrzesien) [inline concent_api/core/views.py]: Same here with timestamp.

File: concent_api/core/views.py (hunk 4 above)
Comment (Jakub89) [reply concent_api/core/views.py]: We can in golem-messages <= 1.6.0, which is currently on master. Should I change it right now, expecting that I will rebase this branch on other that handle newer version of that library?

File: concent_api/core/views.py [hunk 8]
```diff
@@ -374,6 +457,24 @@ def handle_receive_force_get_task_result_upload_for_requestor(decoded_message: m
     decoded_message.sig = None
     return decoded_message
 
+def handle_receive_force_subtasks_results(decoded_message: message.concents.ForceSubtaskResults):
+    assert decoded_message.TYPE in message.registered_message_types
+
+    current_time = int(datetime.datetime.now().timestamp())
+    requestor_force_subtask_results = message.concents.ForceSubtaskResults()
```
Comment (rwrzesien) [inline concent_api/core/views.py]: Please fix styling.

File: concent_api/core/views.py (hunk 4 above)
Comment (rwrzesien) [reply concent_api/core/views.py]: In this case yes, we will be updating because we need messages from > 1.6.0.

File: concent_api/core/views.py (hunk 3 above)
Comment (Jakub89) [reply concent_api/core/views.py]: :+1:

File: concent_api/core/views.py (hunk 6 above)
Comment (Jakub89) [reply concent_api/core/views.py]: :+1:

File: concent_api/core/views.py (hunk 8 above)
Comment (Jakub89) [reply concent_api/core/views.py]: :+1:

File: concent_api/core/tests/test_integration_accept_or_reject_results.py [hunk 9]
```diff
@@ -0,0 +1,428 @@
...
+from core.models            import Message
+from core.models            import ReceiveStatus
+from core.tests.utils       import ConcentIntegrationTestCase
+from utils.testing_helpers  import generate_ecc_key_pair
+
+
+(CONCENT_PRIVATE_KEY, CONCENT_PUBLIC_KEY) = generate_ecc_key_pair()
+
+
+def get_provier_account_status_true_mock(_):
```
Comment (rwrzesien) [inline concent_api/core/tests/test_integration_accept_or_reject_results.py]: provider ;)

File: concent_api/core/tests/test_integration_accept_or_reject_results.py [hunk 10]
```diff
@@ -0,0 +1,428 @@
...
+                    reverse('core:send'),
+                    data                                = serialized_force_subtask_results,
+                    content_type                        = 'application/octet-stream',
+                    HTTP_CONCENT_CLIENT_PUBLIC_KEY      = self._get_encoded_provider_public_key(),
+                )
+
+        self.assertEqual(Message.objects.last().type,               message.concents.ForceSubtaskResults.TYPE)
+        self.assertEqual(Message.objects.last().task_id,            2)
+        self.assertEqual(ReceiveStatus.objects.last().delivered,    False)
+        self.assertEqual(response_1.status_code,                    202)
```
Comment (rwrzesien) [inline concent_api/core/tests/test_integration_accept_or_reject_results.py]: Why not using test_response here too ? This is possible for current implementation.

File: concent_api/core/tests/test_integration_accept_or_reject_results.py [hunk 11]
```diff
@@ -0,0 +1,428 @@
...
+        with mock.patch('core.views.get_provider_account_status', get_provier_account_status_true_mock):
+            with freeze_time("2018-02-05 10:00:30"):
+                response_1 = self.client.post(
+                    reverse('core:send'),
+                    data                                = serialized_force_subtask_results,
+                    content_type                        = 'application/octet-stream',
+                    HTTP_CONCENT_CLIENT_PUBLIC_KEY      = self._get_encoded_provider_public_key(),
+                )
+
+        self.assertEqual(Message.objects.last().type,               message.concents.ForceSubtaskResults.TYPE)
```
Comment (rwrzesien) [inline concent_api/core/tests/test_integration_accept_or_reject_results.py]: Please add method that will check all 3 statements related to Message and ReceiveStatus. It can accept parameters like:
```
message_type, task_id, status_delivered
```

Comment (rwrzesien) [reply concent_api/core/tests/test_integration_accept_or_reject_results.py]: Please add this method for other cases, but here you should use asserts, as we really want to test only seconds response. Btw. method for several asserts would be ok too if it will be repeated (I guess it will).

File: concent_api/core/tests/utils.py [hunk 12]
```diff
@@ -63,5 +82,109 @@ def _get_deserialized_task_to_compute(self, timestamp, deadline, task_id='1'):
...
 
-    def _parse_iso_date_to_timestamp(self, date_string):  # pylint: disable=no-self-use
+    def _get_deserialized_ack_report_computed_task(
+        self,
+        timestamp       = None,
+        deadline        = None,
+        subtask_id      = '1',
+        task_to_compute = None
+    ):
+        """ Returns TaskToCompute deserialized. """
```
Comment (rwrzesien) [inline concent_api/core/tests/utils.py]: Please update comment.

File: concent_api/core/tests/utils.py [hunk 13]
```diff
@@ -63,5 +82,109 @@ def _get_deserialized_task_to_compute(self, timestamp, deadline, task_id='1'):
...
+        timestamp               = None,
+        subtask_id              = 'xxyyxx',
+        payments_ts             = None,
+        requestor_private_key   = None,
+    ):
+        """ Return SubtaskResultsAccepted serialized """
+        with freeze_time(timestamp or self._get_timestamp_string()):
+            subtask_results_accepted = message.concents.SubtaskResultsAccepted(
+                subtask_id      = subtask_id,
+                payments_ts     = self._parse_iso_date_to_timestamp(payments_ts) or
```
Comment (rwrzesien) [inline concent_api/core/tests/utils.py]: Please take a look like it is wrapped in line 138 and do is same way here.

File: concent_api/core/tests/test_integration_accept_or_reject_results.py [hunk 14]
```diff
@@ -0,0 +1,474 @@
...
+        # STEP 1: Provider forces subtask results via Concent.
+        # Request is processed correctly.
+        serialized_force_subtask_results = self._get_serialized_force_subtask_results(
+            timestamp                   = "2018-02-05 10:00:15",
+            ack_report_computed_task    = self._get_deserialized_ack_report_computed_task(
+                subtask_id      = "xxyyzz",
+                task_to_compute = self._get_deserialized_task_to_compute(
+                    timestamp   = "2018-02-05 10:00:00",
+                    deadline    = "2018-02-05 10:00:10",
+                    task_id     = 2,
```
Comment (rwrzesien) [inline concent_api/core/tests/test_integration_accept_or_reject_results.py]: task_id is string now.

File: concent_api/core/tests/test_integration_accept_or_reject_results.py [hunk 15]
```diff
@@ -0,0 +1,474 @@
...
+            status       = 200,
+            key          = self.PROVIDER_PRIVATE_KEY,
+            message_type = message.concents.ServiceRefused,
+            fields       = {
+                'reason':    message.concents.ServiceRefused.REASON.TaskAlreadyProcessed,
+                'timestamp': self._parse_iso_date_to_timestamp("2018-02-05 10:00:31"),
+            }
+        )
+
+    def test_provider_forces_subtask_results_but_doesnt_have_enough_funds_on_this_account(self):
```
Comment (rwrzesien) [inline concent_api/core/tests/test_integration_accept_or_reject_results.py]: Missing `should` in test name.

File: concent_api/core/tests/test_integration_accept_or_reject_results.py [hunk 16]
```diff
@@ -0,0 +1,474 @@
...
+                'timestamp':    self._parse_iso_date_to_timestamp("2018-02-05 11:00:00"),
+            }
+        )
+
+        serialized_force_subtask_results_too_late = self._get_serialized_force_subtask_results(
+            timestamp = "2018-02-05 10:59:59",
+        )
+
+        with mock.patch('core.views.get_provider_account_status', get_provider_account_status_true_mock):
+            with freeze_time("2018-02-05 11:00:00"):
```
Comment (rwrzesien) [inline concent_api/core/tests/test_integration_accept_or_reject_results.py]: Shouldn't it be later, at least 10 seconds ?

File: concent_api/core/views.py [hunk 17]
```diff
@@ -373,6 +406,100 @@ def handle_send_force_get_task_result(request, client_message: message.concents.
...
+    other_party_public_key = decode_other_party_public_key(request)
+
+    if Message.objects.filter(
+        type    = client_message.TYPE,
+        task_id = client_message.ack_report_computed_task.task_to_compute.compute_task_def['task_id']
+    ).exists():
+        return message.concents.ServiceRefused(
+            header = MessageHeader(
+                type_     = message.concents.ServiceRefused.TYPE,
+                timestamp = current_time,
```
Comment (rwrzesien) [inline concent_api/core/views.py]: If current time is used, there is no need to create MessageHeader manually.

File: concent_api/core/views.py [hunk 18]
```diff
@@ -373,6 +406,100 @@ def handle_send_force_get_task_result(request, client_message: message.concents.
...
+                encrypted = False,
+            ),
+            reason      = message.concents.ServiceRefused.REASON.DuplicateRequest,
+        )
+    if not is_provider_account_status_positive(client_message) or tmp_is_provider_account_status_positive(request):
+        return message.concents.ServiceRefused(
+            header = MessageHeader(
+                type_     = message.concents.ServiceRefused.TYPE,
+                timestamp = current_time,
```
Comment (rwrzesien) [inline concent_api/core/views.py]: If current time is used, there is no need to create MessageHeader manually.

File: concent_api/core/views.py [hunk 19]
```diff
@@ -373,6 +406,100 @@ def handle_send_force_get_task_result(request, client_message: message.concents.
...
+            ),
+            reason      = message.concents.ServiceRefused.REASON.TooSmallProviderDeposit,
+        )
+    client_message_send_to_soon = client_message.ack_report_computed_task.timestamp + settings.CONCENT_MESSAGING_TIME
+    client_message_send_to_late = client_message.ack_report_computed_task.timestamp + settings.CONCENT_MESSAGING_TIME + settings.FORCE_ACCEPTANCE_TIME
+    if client_message_send_to_late < current_time:
+        return message.concents.ForceSubtaskResultsRejected(
+            header = MessageHeader(
+                type_     = message.concents.ForceSubtaskResultsRejected.TYPE,
+                timestamp = current_time,
```
Comment (rwrzesien) [inline concent_api/core/views.py]: If current time is used, there is no need to create MessageHeader manually.

File: concent_api/core/views.py [hunk 20]
```diff
@@ -373,6 +406,100 @@ def handle_send_force_get_task_result(request, client_message: message.concents.
...
+                encrypted = False,
+            ),
+            reason = message.concents.ForceSubtaskResultsRejected.REASON.RequestTooLate,
+        )
+    elif current_time > client_message_send_to_soon:
+        return message.concents.ForceSubtaskResultsRejected(
+            header = MessageHeader(
+                type_     = message.concents.ForceSubtaskResultsRejected.TYPE,
+                timestamp = current_time,
```
Comment (rwrzesien) [inline concent_api/core/views.py]: If current time is used, there is no need to create MessageHeader manually.

File: concent_api/core/views.py [hunk 21]
```diff
@@ -373,6 +406,100 @@ def handle_send_force_get_task_result(request, client_message: message.concents.
...
+        return HttpResponse("", status = 202)
+
+
+def handle_send_force_subtask_results_results_response(request, client_message):
+    assert client_message.TYPE in message.registered_message_types
+
+    current_time      = int(datetime.datetime.now().timestamp())
+    client_public_key = decode_client_public_key(request)
+
+    if isinstance(client_message.subtask_results_accepted, message.tasks.SubtaskResultsAccepted):
```
Comment (rwrzesien) [inline concent_api/core/views.py]: `assert` should be added here that `client_message.subtask_results_accepted` contains either `subtask_results_accepted` or `subtask_results_rejected`.

File: concent_api/core/tests/test_integration_accept_or_reject_results.py [hunk 22]
```diff
@@ -0,0 +1,1082 @@
...
+                response_1 = self.client.post(
+                    reverse('core:send'),
+                    data                            = serialized_force_subtask_results_response,
+                    content_type                    = 'application/octet-stream',
+                    HTTP_CONCENT_CLIENT_PUBLIC_KEY  = self._get_encoded_requestor_public_key(),
+                )
+
+        self._test_400_response(response_1)
+
+    def test_requestor_sends_subtask_results_accepted_and_concent_should_return_it_to_provider(self):
```
Comment (rwrzesien) [inline concent_api/core/tests/test_integration_accept_or_reject_results.py]: This test is missing step when Requestor received `ForceSubtaskResults`. But actually it discovers bug or an issue - should it be possible to send `SubtaskResultsAccepted` by requestor before receiving `ForceSubtaskResults` ? I think not.

File: concent_api/core/views.py [hunk 23]
```diff
@@ -112,10 +118,18 @@ def receive(request, _message):
             current_time <= decoded_message_data.task_to_compute.compute_task_def['deadline'] + 2 * settings.CONCENT_MESSAGING_TIME and
             client_public_key == last_undelivered_message_status.message.auth.provider_public_key_bytes
         ):
+            set_message_as_delivered(last_undelivered_message_status)
             decoded_message_data.sig = None
             return decoded_message_data
         return None
 
+    if isinstance(decoded_message_data, message.AckReportComputedTask):
```
Comment (rwrzesien) [inline concent_api/core/views.py]: @Jakub89 This should be moved to branch with "return 204".

File: concent_api/core/views.py [hunk 24]
```diff
@@ -112,10 +118,18 @@ def receive(request, _message):
             current_time <= decoded_message_data.task_to_compute.compute_task_def['deadline'] + 2 * settings.CONCENT_MESSAGING_TIME and
             client_public_key == last_undelivered_message_status.message.auth.provider_public_key_bytes
         ):
+            set_message_as_delivered(last_undelivered_message_status)
```
Comment (rwrzesien) [inline concent_api/core/views.py]: @Jakub89 This should be moved to branch with "return 204".

File: concent_api/core/views.py [hunk 25]
```diff
@@ -157,8 +191,7 @@ def receive_out_of_band(request, _message):
     last_undelivered_receive_status             = Message.objects.filter(
         auth__requestor_public_key = request.META['HTTP_CONCENT_CLIENT_PUBLIC_KEY'],
     ).order_by('timestamp').last()
-
-    current_time      = int(datetime.datetime.now().timestamp())
+    current_time = int(datetime.datetime.now().timestamp())
```
Comment (rwrzesien) [inline concent_api/core/views.py]: Please remove this change.

File: concent_api/core/views.py [hunk 26]
```diff
@@ -373,6 +376,68 @@ def handle_send_force_get_task_result(request, client_message: message.concents.
...
+        if not tmp_is_provider_account_status_positive(request):
+            return message.concents.ServiceRefused(
+                header = MessageHeader(
+                    type_     = message.concents.ServiceRefused.TYPE,
+                    timestamp = current_time,
+                    encrypted = False,
+                ),
+                reason      = message.concents.ServiceRefused.REASON.TooSmallProviderDeposit,
+            )
+        else:
```
Comment (rwrzesien) [inline concent_api/core/views.py]: This `else` is not needed.

Comment (rwrzesien) [inline concent_api/core/views.py]: This `else` is not needed.

File: concent_api/core/views.py [hunk 27]
```diff
@@ -613,6 +618,7 @@ def handle_receive_force_get_task_result_upload_for_requestor(
     force_get_task_result_upload.sig = None
     return force_get_task_result_upload
 
+
```
Comment (rwrzesien) [inline concent_api/core/views.py]: This changes should not belong to this commit.

File: concent_api/core/views.py [hunk 28]
```diff
@@ -869,6 +875,7 @@ def decode_other_party_public_key(request):
 def is_provider_account_status_positive(_message):
     pass
 
+
```
Comment (rwrzesien) [inline concent_api/core/views.py]: This changes should not belong to this commit.

File: concent_api/core/tests/test_integration_accept_or_reject_results.py [hunk 29]
```diff
@@ -0,0 +1,219 @@
...
+from core.models            import Message
+from core.models            import ReceiveStatus
+from core.tests.utils       import ConcentIntegrationTestCase
+from utils.testing_helpers  import generate_ecc_key_pair
+
+
+(CONCENT_PRIVATE_KEY, CONCENT_PUBLIC_KEY) = generate_ecc_key_pair()
+
+
+def get_provider_account_status_true_mock(_):
```
Comment (rwrzesien) [inline concent_api/core/tests/test_integration_accept_or_reject_results.py]: As I have found out, the convention is that we want to name the parameters like this, starting their name with `_`.

File: concent_api/core/tests/test_integration_accept_or_reject_results.py [hunk 30]
```diff
@@ -0,0 +1,219 @@
...
+        # STEP 2: Provider again forces subtask results via Concent with message with the same task_id.
+        # Request is refused.
+        with mock.patch('core.views.is_provider_account_status_positive', get_provider_account_status_true_mock):
+            with freeze_time("2018-02-05 10:00:31"):
+                response_2 = self.client.post(
+                    reverse('core:send'),
+                    data                                = serialized_force_subtask_results,
+                    content_type                        = 'application/octet-stream',
+                    HTTP_CONCENT_CLIENT_PUBLIC_KEY      = self._get_encoded_provider_public_key(),
+                    #HTTP_CONCENT_OTHER_PARTY_PUBLIC_KEY = self._get_encoded_requestor_public_key(),
```
Comment (rwrzesien) [inline concent_api/core/tests/test_integration_accept_or_reject_results.py]: Please uncomment this line now.

File: concent_api/core/views.py [hunk 31]
```diff
@@ -740,3 +805,13 @@ def decode_other_party_public_key(request):
         return decode_key(request.META['HTTP_CONCENT_OTHER_PARTY_PUBLIC_KEY'])
     except binascii.Error:
         raise Http400('The value in the Concent-Other-Party-Public-Key HTTP is not a valid base64-encoded value.')
+
+
+def is_provider_account_status_positive(_message):
+    pass
+
+def tmp_is_provider_account_status_positive(request):
```
Comment (rwrzesien) [inline concent_api/core/views.py]: Please combine it as single function and update logic in view above.

File: concent_api/core/tests/test_integration_accept_or_reject_results.py (hunk 22 above)
Comment (kbeker) [reply concent_api/core/tests/test_integration_accept_or_reject_results.py]: Actually I haven't known that message should be delivered but when Concent will be working with Golem it will be normal to deliver first this message, becuase Requestor won't know which task he will want to accept or reject. But yes, it is bug and we need issue for that

File: concent_api/core/tests/test_integration_accept_or_reject_results.py [hunk 32]
```diff
@@ -348,3 +348,162 @@ def test_requestor_should_not_receive_correct_subtask_results_from_concent_if_as
             }
         )
         self.assertEqual(ReceiveStatus.objects.last().delivered, True)
+
+    def test_requestor_sends_subtask_results_acceptance_but_provider_does_not_submitted_force_subtask_results_concent_should_reject_it(self):
```
Comment (rwrzesien) [inline concent_api/core/tests/test_integration_accept_or_reject_results.py]: Shouldn't it be in `corner cases tests` commit ?

File: concent_api/core/tests/test_integration_accept_or_reject_results.py [hunk 33]
```diff
@@ -348,3 +348,162 @@ def test_requestor_should_not_receive_correct_subtask_results_from_concent_if_as
...
+                response_1 = self.client.post(
+                    reverse('core:send'),
+                    data                            = serialized_force_subtask_results_response,
+                    content_type                    = 'application/octet-stream',
+                    HTTP_CONCENT_CLIENT_PUBLIC_KEY  = self._get_encoded_requestor_public_key(),
+                )
+
+        self._test_400_response(response_1)
+
+    def test_requestor_sends_subtask_results_rejection_but_provider_does_not_submitted_force_subtask_results_concent_should_reject_it(self):
```
Comment (rwrzesien) [inline concent_api/core/tests/test_integration_accept_or_reject_results.py]: Shouldn't it be in 'corner cases tests' commit ?

File: concent_api/core/tests/test_integration_accept_or_reject_results.py [hunk 34]
```diff
@@ -348,3 +348,162 @@ def test_requestor_should_not_receive_correct_subtask_results_from_concent_if_as
...
+                response_1 = self.client.post(
+                    reverse('core:send'),
+                    data                            = serialized_force_subtask_results_response,
+                    content_type                    = 'application/octet-stream',
+                    HTTP_CONCENT_CLIENT_PUBLIC_KEY  = self._get_encoded_requestor_public_key(),
+                )
+
+        self._test_400_response(response_1)
+
+    def test_requestor_sends_subtask_results_accepted_and_concent_should_return_it_to_provider(self):
```
Comment (rwrzesien) [inline concent_api/core/tests/test_integration_accept_or_reject_results.py]: This should be repeated `SubtaskResultsRejected`.

File: concent_api/core/views.py (hunk 24 above)
Comment (Jakub89) [reply concent_api/core/views.py]: I think it should stay, without it We can return that AckReportComputedTask message as many times as we want, as long as timing is right. It is a fix unrelated to usecase, but still needed.
